crates/supervisor/storage/src/providers/derivation_provider.rs

```diff
@@ -233,23 +287,10 @@ where
         &self,
         incoming_pair: DerivedRefPair,
     ) -> Result<(), StorageError> {
-        let mut derived_block_numbers =
-            match self.derived_block_numbers_at_source(incoming_pair.source.number) {
-                Ok(list) => list,
-                Err(StorageError::EntryNotFound(_)) => U64List::default(),
-                Err(err) => {
-                    error!(
-                      target: "supervisor_storage",
-                      incoming_derived_block_pair = %incoming_pair,
-                      %err,
-                      "Failed to get derived block numbers for source block"
-                    );
-                    return Err(err);
-                }
-            };
+        let mut block = self.save_source_block(incoming_pair.source)?;
```

Comment:
I think we shouldn't save the source block here. While saving the derived block, we should assume that `save_source_block` should have been called.